### ISSUE TITLE
unresolved symbol error for macro arg that is used as a let binding

### ISSUE THREAD

**Issue by ikappaki** [clj-kondo/clj-kondo#1935]
**version**

clj-kondo v2022.12.10

**macro usage**

I had look at the below link but  could not locate something relevant to this use case.

See [config.md](https://github.com/clj-kondo/clj-kondo/blob/master/doc/config.md#unrecognized-macros) for related configuration options.

**platform**

native MS-Windows

**problem**

clj-kondo complains about the macro argument being an unresolved symbol but the symbol is used as a let binding name at the macro expanded code.

```issue.clj:8:15: error: Unresolved symbol: b```

**repro**

1. Create an issue.clj file with a macro that uses its first argument as a let binding name, and later using this name at the body of the macro:
```clojure
(ns issue)

(defmacro with-binding
  [binding & body]
  `(let [~binding "something"]
     ~@body))

(with-binding b
  (println :binding b))
```
2. run `clj-kondo --lint issue.clj`, an error is reported:
```
issue.clj:8:15: error: Unresolved symbol: b
linting took 11ms, errors: 1, warnings: 0
```
**expected behavior**

No error should be reported when expanding the macro since the the unresolved symbol is a let binding name in scope after macro expansion.

i.e. the above will print out the binding name when run:
```
$ bb issue.clj
:binding something
```

Thanks

**Comment by borkdude** [comment]
This macro can be used with `:lint-as {issue/with-binding cljs.test/async}` since it has the same syntactical form

**Comment by ikappaki** [comment]
> This macro can be used with `:lint-as {issue/with-binding cljs.test/async}` since it has the same syntactical form

Thanks @borkdude.

I've seen this error reported in clojure-lsp and traced it down to `clj-kondo`. Would you agree that the user expectation here is that macros should be supported as well as functions, or at least should not report false positives, unless it is clearly stated otherwise in the documentation?

I could not find any section in `clojure-lsp` documentation about macros either being unsupported or partially supported, and the `clj-kondo` readme refers to some core macros supported, but does not saying anything about user defined macros:

```
It has support for syntax of commonly used macros like clojure.core.async/alt!!, schema.core/defn and potemkin/import-vars.
```

In addition, this PR template is refereeing to a section about `unrecognized marcos`

```
Unrecognized macros

Clj-kondo only expands a selected set of macros from clojure.core and some popular community libraries. For unrecognized macros you can use these configurations:

    [:lint-as](https://github.com/clj-kondo/clj-kondo/blob/master/doc/config.md#lint-a-custom-macro-like-a-built-in-macro)
    [:unresolved-symbol](https://github.com/clj-kondo/clj-kondo/blob/master/doc/linters.md#unresolved-symbol)
    [:hooks](https://github.com/clj-kondo/clj-kondo/blob/master/doc/config.md#hooks)
```

it is not very clear (at least to me) what it means by `unrecognized macros`. In the example case I posted above, I think the "un/recognized" qualifier does not apply to it. This is a custom macro whose full definition is available to the linter. 

Perhaps there should be a section the main documentation page (ie. readme.md)  about what the user expectation should be with regards to macros in general?  

It appears that the current expectation is that the user has to figure out what the macro linter error is about  for a custom macro, and if it is obviously a false positive try to use some unknown heuristic to `:lint-as` to an unrelated core macro? Not sure how easy this could be for a user that is not familiar with `clj-kondo`'s internals or there is no definite guide to help them with this particular endeavor? 

In my particular case, there are a few macros that use a different number of fixed arguments as let binding names; the above suggestion appears to only be working with one such argument and will fail otherwise.

What is the challenge preventing a linter from correctly linting macros or at least not reporting false positives? From a  naive observer's stand point, linting a macro instance appears to be just a `macroexpand` away .

Thanks

**Comment by borkdude** [comment]
@ikappaki clj-kondo is a static analyzer and can't know about the syntax of custom macros automatically. clj-kondo doesn't execute code, it also doesn't execute `macroexpand` in the context of a JVM.

However, the `:hooks` option does support a `:macroexpand` option that might be closest to what you want.
https://github.com/clj-kondo/clj-kondo/blob/master/doc/hooks.md#macroexpand

You can read more about the challenges with macros here: https://blog.michielborkent.nl/clj-kondo-hooks.html

**Comment by borkdude** [comment]
If you're interested in a Clojure linter that loads your code and does macroexpansion, you can look at eastwood btw.

**Comment by ikappaki** [comment]
Hi @borkdude,

thanks for all the info above, it clarifies a lot.

My use case is the following, which I consider as being a common, not a corner case.

I have an application codebase using a lots of custom macros, most of the macros are define in `test` but there are a few defined in `src` as well. I am using `clojure-lsp` while developing the codebase in the editor and `clojure-lsp diagnostics` for linting the codebase as a whole. Because **I haven't been told or read otherwise**, I am expecting the linters to be reliable and cover all aspect of Clojure. I'm currently getting lint errors reported for the vast majority of the macros; upon further inspection it is obvious that the macros are fine, but the linter is reporting false positives. I know `clojure-lsp` is using `clj-kondo` to do most of its linting work.

IMO there is an issue with how things are currently stand

1. It is not at all clear by reading clj-kondo's `README.md` (or even the other doc pages) that macros are only partially supported, and are likely to result in false positive errors. This most likely to leave the users in a confused state (e.g. I was convinced up until we had this analysis that this was a bug).

Do you think it might be a good idea to update the relevant section in README.md to address the linting expectation for macros? If so, I can do a PR like the following (although I think it will be best if you write something up):

---  
clj-kondo is a static analyzer and can't know about the syntax of custom macros automatically. clj-kondo doesn't execute your project's code, it also doesn't execute macroexpand in the context of a JVM or SCI. Thus, it can't reliably lint macros and might result in false positive errors. See [blog entry on clj-kondo-hooks](https://blog.michielborkent.nl/clj-kondo-hooks.html) about the challenges with linting macros.

However, there are user configuration options that can be used to teach clj kondo about the structure of macro or disable the errors altogether. See [unrecognised-macros](https://github.com/clj-kondo/clj-kondo/blob/master/doc/config.md#unrecognized-macros). 

[and perhaps mention Eastwood as a complementary tool to (reliably?) lint macros from the command line?]

---

For my use case, I think I'm going to use a configuration to ignore the `unresolved-symbol` errors for all the test macros (it would be too much work to try to create hooks for each test macro), and going to try to use Eastwood as a complementary linter on top of `clojure-lsp diagnostics` to cover macro linting out of the box.

**Comment by NoahTheDuke** [comment]
> It is not at all clear by reading clj-kondo's README.md (or even the other doc pages) that macros are only partially supported, and are likely to result in false positive errors. This most likely to leave the users in a confused state (e.g. I was convinced up until we had this analysis that this was a bug).

The README says, "It has support for syntax of commonly used macros like clojure.core.async/alt!!, schema.core/defn and potemkin/import-vars."

And the Configuration doc has the heading "Unrecognized macros" that says, "Clj-kondo only expands a selected set of macros from clojure.core and some popular community libraries." and then links to the relevant parts of the docs for handling macros (lint-as and hooks).

I'm not sure how much more explicit the documentation can be.

**Comment by ikappaki** [comment]
Hi @NoahTheDuke,
> The README says, "It has support for syntax of commonly used macros like clojure.core.async/alt!!, schema.core/defn and potemkin/import-vars."
> 
> And the Configuration doc has the heading "Unrecognized macros" that says, "Clj-kondo only expands a selected set of macros from clojure.core and some popular community libraries." and then links to the relevant parts of the docs for handling user-defined macros (lint-as and hooks).
> 

Please see my second comment in this thread, I've already addressed these two paragraphs.

> I'm not sure how much more explicit the documentation can be. Your suggestion isn't bad but I don't think you can reasonably claim the documentation doesn't make this clear.

I gave an example above of how the documentation might be improved targeting casual users that are not aware of `clj-kondo` macro limitations. It's fine if you think the documentation is clear as it is; it wasn't for me until I had to go through this whole iteration.

Thanks

**Comment by borkdude** [comment]
@ikappaki Not sure if I understood the SCI part of your summary. Clj-kondo does execute macros in a SCI environment, that is the whole point of hooks!

**Comment by borkdude** [comment]
There are reasons why clj-kondo doesn't  execute the macros from your code directly but using the `:macroexpand` option in hooks, you can copy your macros usually directly and then it'll work.

**Comment by borkdude** [comment]
Added something to the README here: https://github.com/clj-kondo/clj-kondo/commit/f6d4c7e262d7577b6221db5f5b41441801b84dec

**Comment by ikappaki** [comment]
> @ikappaki Not sure if I understood the SCI part of your summary. Clj-kondo does execute macros in a SCI environment, that is the whole point of hooks!

Please ignore the mention on SCI (i.e. the "or SCI" part). I was trying to enforce the point that by default `clj-kondo` does not expand macros either on the JVM or even on SCI.  In my mind, before having this whole discussion, I was expecting that custom macros were seamlessly supported, and that seamless support was based on SCI, and thus the mention of SCI in the above. 

> There are reasons why clj-kondo doesn't execute the macros from your code directly but using the `:macroexpand` option in hooks, you can copy your macros usually directly and then it'll work.

I think after reading the blog post, I gained a good understanding why this is the case, and I appreciate all the great work that has been done with hooks to work around the clj-rewrite limitations to support macros.

I am happy with how things were explained to me above and I now know how I should proceed with dealing with the false positive errors I get in my codebase.

The only other thing I'm trying to do here is suggest an improvement to the README.md documentation to set the user expectation about (custom) macro support in `clj-kondo` as I have outlined above in my previous post, if you think is worthwhile. I find the current mention of macros in README.md lacking in coverage and has not helped my case in identifying the tools limitations before hand, which would have saved us this whole discussion.

Thanks

**Comment by borkdude** [comment]
> I find the current mention of macros in README.md lacking in coverage 

Even the latest commit I made 7 minutes ago?

**Comment by ikappaki** [comment]
> Added something to the README here: [f6d4c7e](https://github.com/clj-kondo/clj-kondo/commit/f6d4c7e262d7577b6221db5f5b41441801b84dec)

Thanks looks great. Would it be also possible to mention https://blog.michielborkent.nl/clj-kondo-hooks.html somewhere along the way? This was an eye opener for me and an otherwise very interesting read.

**Comment by borkdude** [comment]
done

**Comment by ikappaki** [comment]
> done

sorry to make you do some work so late on a Sunday evening 😢 it misses the closing parenthesis and points to the hooks doc instead of the blog post  😭 

```
(also see this
[blogpost](https://github.com/clj-kondo/clj-kondo/blob/master/doc/hooks.md).
```

**Comment by borkdude** [comment]
fixed
